src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,144 @@
+
+    protected BitbucketSCMFileSystem(BitbucketApi api, String ref, SCMRevision rev) throws IOException {
+        super(rev);
+        this.ref = ref;
+        this.api = api;
+    }
+
+    @Override
+    public long lastModified() throws IOException {
+        // TODO figure out how to implement this
+        return 0L;
+    }
+
+    @NonNull
+    @Override
+    public SCMFile getRoot() {
+        return new BitbucketSCMFile(this, api, ref);
+    }
+
+    @Extension
+    public static class BuilderImpl extends SCMFileSystem.Builder {
+
+        @Override
+        public boolean supports(SCM source) {
+            return false;
+        }
+
+        @Override
+        public boolean supports(SCMSource source) {
+            return source instanceof BitbucketSCMSource;
+        }
+
+        @Override
+        public SCMFileSystem build(@NonNull Item owner, @NonNull SCM scm, @CheckForNull SCMRevision rev) {
+            return null;
+        }
+        
+        private static StandardUsernamePasswordCredentials lookupScanCredentials(@CheckForNull Item context,
+                @CheckForNull String scanCredentialsId) {
+			if (Util.fixEmpty(scanCredentialsId) == null) {
+				return null;
+			} else {
+				return CredentialsMatchers.firstOrNull(
+						CredentialsProvider.lookupCredentials(
+								StandardUsernamePasswordCredentials.class,
+								context,
+								context instanceof Queue.Task
+								? Tasks.getDefaultAuthenticationOf((Queue.Task) context)
+										: ACL.SYSTEM
+								),
+								CredentialsMatchers.withId(scanCredentialsId)
+						);
+			}
+		}
+
+        @Override
+        public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head, @CheckForNull SCMRevision rev)
+                throws IOException, InterruptedException {
+            BitbucketSCMSource src = (BitbucketSCMSource) source;
+            
+            String credentialsId = src.getCredentialsId();
+            String owner = src.getRepoOwner();
+            String repository = src.getRepository();
+            String serverUrl = src.getServerUrl();
+            StandardUsernamePasswordCredentials credentials;
+            credentials = lookupScanCredentials(src.getOwner(), credentialsId);
+            
+            BitbucketApi apiClient = BitbucketApiFactory.newInstance(serverUrl, credentials, owner, repository);
+           		
+            String ref;
+            if (head instanceof BranchSCMHead) {
+                ref = head.getName();
+            } else if (head instanceof PullRequestSCMHead) {
+                PullRequestSCMHead pr = (PullRequestSCMHead) head;
+                if (!(pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) && pr.getRepository() != null) {
+                    return new BitbucketSCMFileSystem(apiClient, pr.getOriginName(), rev);
+                }
+                // we need to release here as we are not throwing an exception or transferring responsibility to FS
```

Comment:
Looks like left-over coot&paste from GitHub (which needs to track the connection because of strict api rate limits)

Reply:
Copy not coot 

Ffs GitHub won’t let me edit comments on the phobe

Reply:
Done